File: backend/app/schemas/pedidos.py
# -*- coding: utf-8 -*-
"""Schemas de validação/normalização para inputs de pedidos."""
from typing import Any, Optional

from pydantic import AliasChoices, BaseModel, ConfigDict, Field, field_validator


class PedidoBaseSchema(BaseModel):
    """Campos comuns entre criação e edição."""

    model_config = ConfigDict(extra="ignore", populate_by_name=True)

    cliente: Optional[str] = None
    telefone_cliente: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("telefone_cliente", "telefone"),
    )
    destinatario: Optional[str] = None
    tipo_pedido: Optional[str] = "Entrega"
    fonte_pedido_id: Any = None
    fonte_pedido: Optional[str] = None
    produto: Optional[str] = None
    flores_cor: Optional[str] = None
    valor: Optional[str] = None
    horario: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("horario", "hora_entrega"),
    )
    dia_entrega: Optional[str] = Field(
        default=None,
        validation_alias=AliasChoices("dia_entrega", "data_entrega"),
    )
    cep: Optional[str] = None
    rua: Optional[str] = None
    numero: Optional[str] = None
    bairro: Optional[str] = None
    cidade: Optional[str] = None
    endereco: Optional[str] = None
    obs_entrega: Optional[str] = None
    mensagem: Optional[str] = None
    pagamento: Optional[str] = None
    observacoes: Optional[str] = None
    status_pagamento: Optional[str] = None
    status: Optional[str] = None
    quantidade: Any = 1
    cliente_id: Any = None

    @field_validator(
        "cliente",
        "telefone_cliente",
        "destinatario",
        "tipo_pedido",
        "fonte_pedido",
        "produto",
        "flores_cor",
        "valor",
        "horario",
        "dia_entrega",
        "cep",
        "rua",
        "numero",
        "bairro",
        "cidade",
        "endereco",
        "obs_entrega",
        "mensagem",
        "pagamento",
        "observacoes",
        "status_pagamento",
        "status",
        mode="before",
    )
    @classmethod
    def strip_strings(cls, value: Any) -> Optional[str]:
        if value is None:
            return None
        if isinstance(value, str):
            return value.strip()
        return str(value).strip()


class PedidoCreateSchema(PedidoBaseSchema):
    """Schema para criação de pedido."""


class PedidoUpdateSchema(PedidoBaseSchema):
    """Schema para edição de pedido."""

File: frontend_v2/src/api/endpoints/pedidos.ts
/**
 * Orders API endpoints
 * All API calls via React Query hooks (no manual useEffect)
 */

import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { createApiRequest } from '../http';
import { useAuth } from '../../features/auth/authStore';

// Types
export interface Pedido {
  id: number;
  cliente: string;
  telefone_cliente: string;
  destinatario: string;
  tipo_pedido: 'Entrega' | 'Retirada';
  produto: string;
  flores_cor?: string;
  valor?: string;
  dia_entrega: string; // YYYY-MM-DD
  horario: string;
  cep?: string;
  rua?: string;
  numero?: string;
  bairro?: string;
  cidade?: string;
  endereco?: string;
  obs_entrega?: string;
  mensagem?: string;
  pagamento?: string;
  observacoes?: string;
  fonte_pedido?: string;
  fonte_pedido_id?: number;
  fonte_pedido_nome?: string;
  status_pagamento?: string;
  status: string;
  quantidade: number;
  oculto: boolean;
  impresso: boolean;
  cliente_id?: number;
  distancia_km?: number;
  taxa_entrega?: number;
  coords_lat?: number;
  coords_lon?: number;
  created_at?: string;
  updated_at?: string;
}

export interface PedidosResponse {
  success: boolean;
  pedidos: Pedido[];
  total: number;
}

export interface PedidosFilters {
  status?: string;
  data_inicio?: string; // YYYY-MM-DD
  data_fim?: string; // YYYY-MM-DD
  search?: string;
}

/**
 * Get orders with filters
 */
export function usePedidos(filters: PedidosFilters = {}) {
  const { getAuthHeader } = useAuth();
  const apiRequest = createApiRequest(getAuthHeader);

  return useQuery<PedidosResponse>({
    queryKey: ['pedidos', filters],
    queryFn: async () => {
      const params = new URLSearchParams();
      if (filters.status) params.append('status', filters.status);
      if (filters.data_inicio) params.append('data_inicio', filters.data_inicio);
      if (filters.data_fim) params.append('data_fim', filters.data_fim);
      if (filters.search) params.append('search', filters.search);

      const queryString = params.toString();
      const endpoint = `/pedidos${queryString ? `?${queryString}` : ''}`;

      const response = await apiRequest<PedidosResponse>(endpoint);
      if (!response.ok) {
        throw new Error(response.message);
      }
      return response.data;
    },
    staleTime: 5000, // 5 seconds
    refetchInterval: 15000, // 15 seconds
    refetchOnWindowFocus: true,
    keepPreviousData: true, // Maintains previous data when filters change
  });
}

/**
 * Get single order by ID
 */
export function usePedido(id: number) {
  const { getAuthHeader } = useAuth();
  const apiRequest = createApiRequest(getAuthHeader);

  return useQuery<{ success: boolean; pedido: Pedido }>({
    queryKey: ['pedido', id],
    queryFn: async () => {
      const response = await apiRequest<{ success: boolean; pedido: Pedido }>(`/pedidos/${id}`);
      if (!response.ok) {
        throw new Error(response.message);
      }
      return response.data;
    },
    enabled: !!id,
  });
}

